README.md
`Buzzwords: #typescript #angularjs #immutablejs #gulp #less #bower #selenium #karma #protractor`



beep
boop
meep

[![Build Status](https://magnum.travis-ci.com/NCI-GDC/portal-ui.svg?token=uvYjzX9Pbq2o3AVCP1S4&branch=master)](https://magnum.travis-ci.com/NCI-GDC/portal-ui)

- [Technologies](#technologies)
- [Installation](#installation)
- [Tests](#tests)
- [Development](#development)
- [Contributing](#contributing)
- [Production](#production)
- [Verifying Tags](#verify)
- [Resources](#resources)

# Technologies

- [Typescript](http://www.typescriptlang.org/) - TypeScript lets you write JavaScript the way you really want to.
- [Angular](https://angularjs.org/) - HTML enhanced for web apps
- [AngularUI](http://angular-ui.github.io/) - The companion suite(s) to the AngularJS framework.
- [ImmutableJS](https://github.com/facebook/immutable-js) - Immutable Data Collections for Javascript
- [Gulp.js](http://gulpjs.com/) - The streaming build system
- [Karma](http://karma-runner.github.io/0.12/index.html) - Spectacular Test Runner for Javascript
- [Protractor](https://github.com/angular/protractor) - E2E test framework for Angular apps
- [Less](http://learnboost.github.io/stylus/)

# Installation

## Global Dependencies

Before continuing you must have the following programs installed:

- [Node](http://nodejs.org/)

## Setup Script

Running the setup script will:

1. Setup the needed git hook for the project
2. Install npm and bower dependencies

```
❯ ./setup.sh
commit-msg already exists! Backing up to commit-msg.bak...  OK
Setting up commit-msg git hook...                           OK
Making commit-msg executable...                             OK
Making validate-commit.py executable...                     OK
Confirming Node is installed...                             OK
Installing NPM dependencies...
...
Setup Successful!

```

### Running Locally
#### Remote API & ElasticSearch, Local UI
Connect to VPN and run ui with
```
GDC_API=https://gdc-portal.nci.nih.gov/auth/api/v0 GDC_FAKE_AUTH=true npm start
```

#### Local API, Remote ElastiSearch and Local UI
Connect to VPN and run api with
```
GDC_ES_INDEX=gdc_from_graph GDC_ES_HOST=ip GDC_ES_USER=user GDC_ES_PASS=pw GDC_PORTAL_ENDPOINT=http://localhost:3000 GDC_FAKE_DOWNLOAD=True python run.py
```
ES creds can be found on any API machine in `/var/www/gdcapi/gdcapi.wsgi`

UI can simply be run as usual with
```
npm start
```

#### Local API, ElasticSearch and UI
Not recommended, would require loading local ES with data.

### Authentication
In order to properly run the UI and login to test the auth you will need to run the application
in a specific way. This includes `sudo` and running the app on port `80`. You will need to ensure
that you don't have something else using that port (like a local apache setup).

The following command should work:
`sudo GDC_API=http://portal.gdc.nci.nih.gov:5000 GDC_PORT=80 npm start`

#### Modifying /etc/hosts
In order to support local use of the login system we need to add the following
to your `/etc/hosts` file:

`127.0.0.1 gdc-portal.nci.nih.gov`

### ElasticSearch
If you are connecting to a local ES, edit path-to-elastic-search/config/elasticsearch.yml, find the line with http.max_content_length, add
```
http.max_initial_line_length: 1000mb
```
to increase the max length of a HTTP URL

### Git hooks

Git commit-msg hook is based on Angular's [Guidelines](https://docs.google.com/document/d/1QrDFcIiPjSLDn3EL15IJygNPiHORgU1_OOAqWjiDU5Y/edit#)

The git hooks can be found in `git_hooks/`

### Project Dependencies

Project dependencies are managed using [Bower](http://bower.io/) and [NPM](https://www.npmjs.org/)

# Tests

Unit tests are run using Karma

```
❯ npm test
...
[16:44:06] Starting Karma server...
WARN [karma]: Port 9876 in use
INFO [karma]: Karma v0.12.24 server started at http://localhost:9877/
INFO [launcher]: Starting browser PhantomJS
INFO [PhantomJS 1.9.7 (Mac OS X)]: Connected on socket m-mK9_Udpg1Jb226Ws3o with id 36984552
...
Finished in 0.014 secs / 0.001 secs
```

Browser tests are run with Protractor

```
❯ npm run prot
...
[16:34:28] Starting 'webdriver'...
selenium standalone is up to date.
chromedriver is up to date.
[16:34:28] Finished 'webdriver' after 199 ms
[16:34:28] Starting 'protractor'...
Using ChromeDriver directly...
...
[16:34:32] Finished 'protractor' after 4.23 s

```

# Development

The development server is setup using Browsersync

```
❯ GDC_API="http://gdc-portal.nci.nih.gov:5000" npm start
[16:47:02] Environment Development
...
[BS] Local URL: http://localhost:3000
[BS] External URL: http://gdc-portal.nci.nih.gov:3000
[BS] Serving files from: dist
```

# Contributing

Read how to contribute [here](https://github.com/NCI-GDC/portal-ui/blob/master/CONTRIBUTING.md)

# Production

Gulp bundles all the assets in production mode and creates unique file names for caching

```
❯ npm dist
...
Found 3 matching tests...
...
Environment Production
...
Starting selenium server in parallel mode...
...
❯ cd dist
❯ python -m SimpleHTTPServer
Serving HTTP on 0.0.0.0 port 8000
...
❯ open http://localhost:8000/
```

# Verifying Tags

```
❯ git show maintainer-pgp-pub | gpg --import
❯ git tag --verify [signed-tag-name]
```

# Resources

- [angularjs-via-typescript-controllers](http://kodeyak.wordpress.com/2014/02/12/angularjs-via-typescript-controllers/)
- [AngularJS + TypeScript : Controllers, Best Practice](https://www.youtube.com/watch?v=WdtVn_8K17E)
- [Angular Services using TypeScript : Best Practices](https://www.youtube.com/watch?v=Yis8m3BdnEM)


CHANGELOG.md
## [0.0.3](https://github.com/nci-gdc/portal-ui}/compare/1.3.0...0.0.3) (2016-10-05)
1

### :package: Refactor

- namespace npm stuff ([6ce6554](https://github.com/nci-gdc/portal-ui/commit/6ce65541b437dc216592ecedb7a77299a9d32a52))
- more js into packages/ for lerna ([9a71451](https://github.com/nci-gdc/portal-ui/commit/9a7145189d73642cdc683f9ba5f584b770acd5af))

<a name"1.3.0"></a>
## 1.3.0 (2016-08-29)


#### Bug Fixes

* **annotation:** PRTL-182 no word break in notes ([b808f199](https://github.com/NCI-GDC/portal-ui/commit/b808f199), closes [PGDC-182](https://jira.opensciencedatacloud.org/browse/PGDC-182))
* **app:** PRTL-237 can't find data banner ([79d4e46b](https://github.com/NCI-GDC/portal-ui/commit/79d4e46b))
* **cart:** PRTL-181 sync across tabs ([dad250d6](https://github.com/NCI-GDC/portal-ui/commit/dad250d6))
* **customFacets:** PRTL-184 use actual checkbox ([8dac6140](https://github.com/NCI-GDC/portal-ui/commit/8dac6140), closes [PGDC-184](https://jira.opensciencedatacloud.org/browse/PGDC-184))
* **deploy:** re-add prestart/prebuild ([8be9b418](https://github.com/NCI-GDC/portal-ui/commit/8be9b418))
* **downloader:** do not send bam regions as csv ([ff186b71](https://github.com/NCI-GDC/portal-ui/commit/ff186b71))
* **facet:** calc width facet sidebar ([a7916583](https://github.com/NCI-GDC/portal-ui/commit/a7916583))
* **facets:** 508 label name correction ([98f4acdd](https://github.com/NCI-GDC/portal-ui/commit/98f4acdd))
* **file:** PRTL-86 hide padding if 0 assoc cases ([81e3164f](https://github.com/NCI-GDC/portal-ui/commit/81e3164f), closes [PGDC-86](https://jira.opensciencedatacloud.org/browse/PGDC-86))
* **gql:**
  * correct toBytes equation ([b059c8f2](https://github.com/NCI-GDC/portal-ui/commit/b059c8f2))
  * PRTL-123 PRTL-183 blur click outside,esc ([699245fd](https://github.com/NCI-GDC/portal-ui/commit/699245fd), closes [PGDC-123](https://jira.opensciencedatacloud.org/browse/PGDC-123), [PGDC-183](https://jira.opensciencedatacloud.org/browse/PGDC-183))
* **header:**
  * PRTL-179 conserve space in header ([393ffda2](https://github.com/NCI-GDC/portal-ui/commit/393ffda2))
  * improve responsive style ([bc1582b3](https://github.com/NCI-GDC/portal-ui/commit/bc1582b3))
* **quickSearch:** 617 highlight in pane ([9974c165](https://github.com/NCI-GDC/portal-ui/commit/9974c165), closes [PGDC-617](https://jira.opensciencedatacloud.org/browse/PGDC-617))
* **table:**
  * PRTL-96 correct sticky th ux ([1673d55e](https://github.com/NCI-GDC/portal-ui/commit/1673d55e))
  * wrap text in annotation category cell ([1ad08aab](https://github.com/NCI-GDC/portal-ui/commit/1ad08aab))


#### Features

* **cart:** [PRTL-177] combine metadata downloads ([d3bb1afa](https://github.com/NCI-GDC/portal-ui/commit/d3bb1afa))
* **download:** 2468 bam slicing dl enhancement ([2d6df6f0](https://github.com/NCI-GDC/portal-ui/commit/2d6df6f0), closes [PGDC-2468](https://jira.opensciencedatacloud.org/browse/PGDC-2468))
* **export:** 1892 restored large-dataset download ([43501b69](https://github.com/NCI-GDC/portal-ui/commit/43501b69), closes [PGDC-1892](https://jira.opensciencedatacloud.org/browse/PGDC-1892))
* **facet:** 2405 open typeahead on input click ([13e9d12f](https://github.com/NCI-GDC/portal-ui/commit/13e9d12f), closes [PGDC-2405](https://jira.opensciencedatacloud.org/browse/PGDC-2405))
* **facets:** PRTL-104 ellipsis for overflow ([137ca40d](https://github.com/NCI-GDC/portal-ui/commit/137ca40d), closes [PGDC-104](https://jira.opensciencedatacloud.org/browse/PGDC-104))
* **file:** PRTL-75 link to source if 1 file ([9b379784](https://github.com/NCI-GDC/portal-ui/commit/9b379784))
* **gql:**
  * support 'b/B' for bytes ([86aa94d4](https://github.com/NCI-GDC/portal-ui/commit/86aa94d4))
  * do not require quotes for file size ([58fcbd2d](https://github.com/NCI-GDC/portal-ui/commit/58fcbd2d))
  * 1778 allows file size unit ([90e181bf](https://github.com/NCI-GDC/portal-ui/commit/90e181bf), closes [PGDC-1778](https://jira.opensciencedatacloud.org/browse/PGDC-1778))
  * 2350 find/report bogus fields. ([2fc71371](https://github.com/NCI-GDC/portal-ui/commit/2fc71371), closes [PGDC-2350](https://jira.opensciencedatacloud.org/browse/PGDC-2350))
* **header:** PRTL-136 login opens on enter ([90dfab09](https://github.com/NCI-GDC/portal-ui/commit/90dfab09))
* **search:** PRTL-147 save facet tab in url state ([ec51fc63](https://github.com/NCI-GDC/portal-ui/commit/ec51fc63))


<a name"1.2.4"></a>
### 1.2.4 (2016-08-12)


#### Bug Fixes

* **app:** PRTL-237 can't find data banner ([b3808e45](https://github.com/NCI-GDC/portal-ui/commit/b3808e45))


<a name"1.2.3"></a>
### 1.2.3 (2016-08-10)


#### Bug Fixes

* **app:**
  * PRTL-237 banner about portal status ([5384c825](https://github.com/NCI-GDC/portal-ui/commit/5384c825))


<a name"1.2.2"></a>
### 1.2.2 (2016-08-09)


#### Bug Fixes

* **app:**
  * remove transition from gdc-body ([2e55f363](https://github.com/NCI-GDC/portal-ui/commit/2e55f363))


#### Features

* **home:** update Data Release date ([bc1ed465](https://github.com/NCI-GDC/portal-ui/commit/bc1ed465))


<a name"1.2.1"></a>
### 1.2.1 (2016-08-08)


#### Bug Fixes

* **app:** PRTL-226 add project info to files ([718d930c](https://github.com/NCI-GDC/portal-ui/commit/718d930c))


#### Features

* **app:** PRTL-172 legacy data banner ([868131d6](https://github.com/NCI-GDC/portal-ui/commit/868131d6))


<a name"1.2.0"></a>
## 1.2.0 (2016-06-30)


#### Bug Fixes

* **app:**
  * 2588 move website icon up ([2e427476](https://github.com/NCI-GDC/portal-ui/commit/2e427476), closes [PGDC-2588](https://jira.opensciencedatacloud.org/browse/PGDC-2588))
  * 2429 previous state when back from 404 ([86b5da29](https://github.com/NCI-GDC/portal-ui/commit/86b5da29), closes [PGDC-2429](https://jira.opensciencedatacloud.org/browse/PGDC-2429))
* **cart:**
  * cannot download cart with over 5gb ([83c40084](https://github.com/NCI-GDC/portal-ui/commit/83c40084))

gulpfile.js
var gulp = require('gulp');
var $ = require('gulp-load-plugins')();
$.ngAnnotate = require('gulp-ng-annotate');
var del = require('del');
var runSequence = require('run-sequence');
var browserSync = require('browser-sync');
var reload = browserSync.reload;
var pagespeed = require('psi');
var fs = require('graceful-fs');
var packageJSON = require("./package.json");
var modRewrite = require('connect-modrewrite');
var mkdirp = require("mkdirp");
var browserify = require('browserify');
var source = require('vinyl-source-stream');
var rename = require('gulp-rename');
var jest = require('jest-cli');
var jestConfig = {
    rootDir: 'app/react-components'
};

var env = {
  api: process.env.GDC_API || "http://localhost:5000",
  auth: process.env.GDC_AUTH || "https://gdc-portal.nci.nih.gov/auth",
  base: process.env.GDC_BASE || "/",
  port: process.env.GDC_PORT || 3000,
  fake_auth: process.env.GDC_FAKE_AUTH || false,
  es_host: process.env.ES_HOST || "http://localhost:9200",
  es_index_version: process.env.ES_INDEX_VERSION || "gdc-r7",
};

var AUTOPREFIXER_BROWSERS = [
  'ie >= 10',
  'ie_mob >= 10',
  'ff >= 30',
  'chrome >= 34',
  'safari >= 7',
  'opera >= 23',
  'ios >= 7',
  'android >= 4.4',
  'bb >= 10'
];

var production = false;// process.env.NODE_ENV === "production";
$.util.log('Environment', $.util.colors.blue(production ? 'Production' : 'Development'));

// <paths>
var paths = {
  "src": "./src",
  "dev": "./dist",
  "stage": "./dist",
  "bower": "./bower_components/**/*.{js,css,map,eot,svg,ttf,woff,woff2}",
  js: {},
  html: {}
};
paths.ddest = production ? paths.stage : paths.dev;
paths.js.src = paths.src + "/app/**/*";
paths.js.dest = paths.ddest + "/js";
paths.vendor = paths.ddest + "/libs";
paths.html.src = paths.src + "/index.html";
paths.html.dest = paths.dest + "/index.html";
// </paths>

gulp.task('logs', function () {
  require('conventional-changelog')({
    repository: packageJSON.repository,
    version: packageJSON.version,
    issueLink: function (id) {
      return '[PGDC-' + id + '](https://jira.opensciencedatacloud.org/browse/PGDC-' + id + ')'
    }
    //,from: "8b8e5953153997e8cf53c68237fadd0e7c0cacdd"
  }, function (err, log) {
    fs.writeFile('CHANGELOG.md', log);
  });
});

gulp.task('clean', del.bind(null, ['.tmp', 'dist']));

gulp.task('i18n:pot', function () {
  return gulp.src(['app/scripts/**/*.html', 'app/scripts/**/*.js'])
      .pipe($.angularGettext.extract('template.pot'))
      .pipe(gulp.dest('translations/'));
});

gulp.task('i18n:translations', ['i18n:pot'], function () {
  return gulp.src('translations/**/*.po')
      .pipe($.angularGettext.compile({
        format: 'javascript',
        defaultLanguage: 'en'
      }))
      .pipe(gulp.dest('dist/translations/'));
});

gulp.task('i18n:build', ['i18n:translations'], function () {
  var f = production ? 'translations.min.js' : 'translations.js';

  return gulp.src('dist/translations/**/*.js')
      .pipe($.concat(f))
      .pipe(gulp.dest('dist/js'));
});

// No file gets built in the case no translations exist.
// Build one to prevent 404s in browser.
gulp.task('i18n:exists', function () {
  var f = production ? "translations.min.js" : "translations.js";

  fs.readFile("dist/js/" + f, function (err) {
    if (err) {
      mkdirp("dist/js", function (err) {
        if (err) {
          throw err;
        }

        fs.writeFile("dist/js/" + f, "");
      });
    }
  });
});

gulp.task('i18n:clean', del.bind(null, ['dist/translations', 'dist/js/translations.js', 'dist/js/translations.min.js']));

gulp.task('i18n', function (cb) {
  runSequence('i18n:clean', ['i18n:build', 'i18n:exists'], cb);
});

// Optimize Images
gulp.task('images', function () {
  return gulp.src('app/images/**/*')
//      .pipe($.imagemin({
//        progressive: true,
//        interlaced: true
//      }))
      .pipe(gulp.dest('dist/images'))
      .pipe($.size({title: 'images'}));
});

gulp.task("config", function () {
  var f = production ? "config.min.js" : "config.js";

  fs.readFile("app/config.js", "UTF-8", function (err, content) {
    if (err) {
      throw err;
    }

    $.git.exec({args: "rev-parse --short HEAD"}, function (err, stdout) {
      if (err) {
        throw err;
      }

      content = content.replace(/__VERSION__/g, packageJSON.version);
      content = content.replace(/__COMMIT__/g, stdout.replace(/[\r\n]/, ""));
      content = content.replace(/__API__/, env.api);
      content = content.replace(/__AUTH__/g, env.auth);
      content = content.replace(/__PRODUCTION__/, production);
      content = content.replace(/__FAKE_AUTH__/, env.fake_auth);
      content = content.replace(/__ES_HOST__/, env.es_host);
      content = content.replace(/__ES_INDEX_VERSION__/, env.es_index_version);

      // Ensures path is in place, as I've had occurances where it may not be.
      mkdirp("dist/js", function (err) {
        if (err) {
          throw err;
        }

        fs.writeFileSync("dist/js/" + f, content);
      });
    });
  });
});

// Copy Web Fonts To Dist
gulp.task('fonts', function () {
  return gulp.src(['app/fonts/**'])
      .pipe(gulp.dest('dist/fonts'))
      .pipe($.size({title: 'fonts'}));
});

// Copy External non Bower Libraries To Dist
gulp.task('vendor', function () {
  return gulp.src(['app/vendor/**'])
      .pipe(gulp.dest('dist/libs'))
      .pipe($.size({title: 'vendor'}));
});

// Compile and Automatically Prefix Stylesheets
gulp.task('styles', function () {
  var f = production ? 'styles.min.css' : 'styles.css';

  return gulp.src('app/styles/app.less')
      .pipe($.changed('styles', {extension: '.less'}))
      .pipe($.less().on('error', $.util.log))
      .pipe($.autoprefixer({browsers: AUTOPREFIXER_BROWSERS}))
      .pipe($.concat(f))
      //.pipe($.if(production, $.csso()))
    // Concatenate And Minify Styles
      .pipe(gulp.dest('dist/css'))
      .pipe($.size({title: 'styles'}));
});

gulp.task('js:bower', function () {
  var stream = gulp.src(paths.bower);

  if (production) {
    var filter = $.filter('**/*.min.{js,css,map}');

    stream
        .pipe(filter)
        .pipe($.foreach(function (stream, file) {
          if (file.relative.indexOf(".js")) {
            var fileName = file.relative.substr(file.relative.lastIndexOf("/") + 1);
            fileName = fileName.substr(0, fileName.lastIndexOf(".min") + 4) + ".js.map";

            var path = file.relative.substr(0, file.relative.lastIndexOf("/") + 1) + fileName;
            return stream
                .pipe($.replace(fileName, "/libs/" + path));
          }

          return stream;
        }))
        .pipe($.rev())
        .pipe(gulp.dest(paths.vendor))
        .pipe($.rev.manifest())
        .pipe(gulp.dest(paths.vendor))
        .pipe(filter.restore());
  }

  return stream.pipe(gulp.dest('dist/libs'))
      .pipe($.size({title: 'js:bower'}));
});

// Scan Your HTML For Assets & Optimize Them
gulp.task('rev', ['html'], function () {
  var stream = gulp.src('dist/index.html');

  if (production) {
    var manifest = paths.vendor + "/rev-manifest.json";
    var vendorFiles = fs.existsSync(manifest) ? require(manifest) : [];
    var assets = $.useref.assets({searchPath: 'dist'});

    for (var file in vendorFiles) {
      if (vendorFiles.hasOwnProperty(file)) {
        stream = stream.pipe($.replace(file, vendorFiles[file]));
      }
    }

    stream.pipe(assets)
        .pipe($.rev())
        .pipe(assets.restore())
        .pipe($.useref())
        .pipe($.revReplace())
        .pipe(gulp.dest('dist'))
        .pipe($.gzip())
        .pipe(gulp.dest('dist'));
  }

  return stream
      .pipe(gulp.dest('dist'))
      .pipe($.size({title: 'rev'}));
});

// Scan Your HTML For Assets & Optimize Them
gulp.task('html', ['js:bower', 'ng:templates'], function () {
  var stream = gulp.src('app/index.html');

  if (production) {
    stream
        .pipe($.replace('.js', '.min.js'))
        .pipe($.replace('d3-tip/index.min.js', 'd3-tip/index.js'))
        .pipe($.replace('moment/moment.min.js', 'moment/min/moment.min.js'))
        .pipe($.replace('analytics.min.js', 'analytics.js'))
        .pipe($.replace('FileSaver.min.min.js', 'FileSaver.min.js'))
        .pipe($.replace('.css', '.min.css'))
        .pipe($.replace('src/css/bootcards-desktop.min.css', 'src/css/bootcards-desktop.css'))
        .pipe($.replace('ngprogress-lite.min.css', 'ngprogress-lite.css'));
  }
  return stream
      .pipe($.replace('__BASE__', env.base))
      .pipe(gulp.dest('dist'))
      .pipe($.size({title: 'html'}));
});

// <tests>
gulp.task('test', ['clean', 'ts:compile', 'ng:templates', 'jest'], function () {
  runSequence('karma:once')
});

gulp.task('jest', function() {
  jest.runCLI({ config : jestConfig }, ".");
});

gulp.task('plato', function () {
  return gulp.src('.tmp/scripts/**/*.js')
      .pipe($.plato('report'));
});

gulp.task('karma:once', function () {
  // Be sure to return the stream
  return gulp.src('path') // not a real path
      .pipe($.karma({
        configFile: 'karma.conf.js',
        action: 'run'
      }))
      .on('error', function (err) {
        // Make sure failed tests cause gulp to exit non-zero
        throw err;
      });
});

gulp.task('karma:watch', function () {
  return gulp.src('path') // not a real path
      .pipe($.karma({
        configFile: 'karma.conf.js',
        action: 'watch'
      }));
});

gulp.task('webdriver', $.protractor.webdriver_update);

gulp.task('protractor', ['webdriver'], function () {
  return gulp.src('app/tests/**/*.spec.js')
      .pipe($.protractor.protractor({configFile: 'protractor.conf.js'}));
});
// </tests>

// <typescript>
var tsProject = $.typescript.createProject({
    "target": "ES3",
		"module": "commonjs",
		"declaration": true,
		"noExternalResolve": false
});

gulp.task('ts:compile', function () {
  var f = production ? 'app.min.js' : 'app.js';
  var tsResult = gulp.src('app/**/*.ts')
      .pipe($.typescript(tsProject));

  tsResult.dts.pipe(gulp.dest('dist/dts'));
  tsResult.js.pipe(gulp.dest('.tmp'));

  return tsResult.js
      .pipe($.sourcemaps.init())
      .pipe($.concat(f))
      .pipe($.ngAnnotate())
      .pipe($.if(production, $.uglify()))
      .pipe($.wrap({src: './iife.txt'}))
      .pipe($.sourcemaps.write())
      .pipe(gulp.dest('dist/js'))
      .pipe($.size({title: 'typescript'}));
});
// </typescript>

gulp.task('babel', function() {
  return browserify('app/react-components/index.js')
    .transform("babelify", {presets: ["react", "es2015", "stage-0"]})
    .transform('browserify-css', { autoInject: true })
    .bundle()
    .on('error', function (err) {
        console.log(err.message);
        console.log(err.stack);
        this.emit("end");
    })
    .pipe(source('react-components.js'))
    .pipe(gulp.dest('./dist/js'));
});

gulp.task('vis', function () {
  return gulp.src(['app/scripts/humanbody.js'])
    .pipe(gulp.dest('dist/js'))
})

// <ng-templates>
gulp.task('ng:templates', function () {
  var f = production ? 'templates.min.js' : 'templates.js';

  return gulp.src('app/scripts/**/templates/*.html')
      .pipe($.minifyHtml({
        empty: true,
        spare: true,
        quotes: true
      }))
      .pipe($.ngHtml2js({
        moduleName: function (file) {
          var path = file.path.split('/'),
              folder = path[path.length - 2];
          return folder.replace(/-[a-z]/g, function (match) {
            return match.substr(1).toUpperCase() + 'templates';
          });
        }
      }))
      .pipe($.concat(f))
      .pipe($.if(production, $.uglify()))
      .pipe(gulp.dest('dist/js'))
      .pipe($.size({title: 'ng:templates'}));
});
// </ng-templates>

// Watch Files For Changes & Reload
gulp.task('serve:web', function (cb) {
  var bsOpts = {
    open: false,
    notify: false,
    // Run as an https by uncommenting 'https: true'
    // Note: this uses an unsigned certificate which on first access
    //       will present a certificate warning in the browser.
    // https: true,
    server: {
      middleware: [
        modRewrite(['!\\.html|\\images|\\.js|\\.css|\\.png|\\.jpg|\\.woff|\\.ttf|\\.svg /index.html [L]'])
      ],
      baseDir: 'dist'
    },
    port: env.port,
    host: "gdc-portal.nci.nih.gov"
//    open: "external"
  };
//    bsOpts.tunnel = production ? 'oicrgdcdev' : false;

  browserSync(bsOpts);

  if (!production) {
    gulp.watch(['app/**/*.html'], ['html', reload]);
    gulp.watch(['app/**/*.{less,css}'], ['styles', reload]);
    gulp.watch(['app/scripts/**/*.ts'], ['ts:compile', reload]);
    gulp.watch(['app/scripts/**/*.html'], ['ng:templates', reload]);
    gulp.watch(['app/images/**/*'], ['images', reload]);
    gulp.watch(['app/react-components/**/*.js'], ['babel', 'jest', reload]);
  }
});

gulp.task('pegjs', function () {
  var PEG = require('pegjs');
  var input = fs.readFileSync("gql.pegjs", {encoding: 'utf8'});
  var parser = PEG.buildParser(input, {output: "source"});
  mkdirp("dist/js", function (err) {
    if (err) {
      throw err;
    }
    var out = production ? "dist/js/gql.min.js" : "dist/js/gql.js";
    fs.writeFileSync(out, "window.gql = " + parser);
    if (production) {
      gulp.src(out)
        .pipe($.uglify())
        .pipe(gulp.dest('dist/js'))
        .pipe($.size({title: 'GQL'}));
    }

  });
});

gulp.task('serve', function (cb) {
  runSequence('default', ['karma:watch',
   'serve:web'], cb);
});

// Build Production Files, the Default Task
gulp.task('default', ['clean'], function (cb) {
  runSequence('styles', [
    'rev',
    'images',
    'fonts',
    'vendor',
    'ts:compile',
    'babel',
    'i18n',
    'config',
    'pegjs',
    'vis'
  ], cb);
});

// Run PageSpeed Insights
// Update `url` below to the public URL for your site
gulp.task('pagespeed', pagespeed.bind(null, {
  // By default, we use the PageSpeed Insights
  // free (no API key) tier. You can use a Google
  // Developer API key if you have one. See
  // http://goo.gl/RkN0vE for info key: 'YOUR_API_KEY'
  url: 'https://oicrgdcdev.localtunnel.me',
  strategy: 'mobile'
}));

// Load custom tasks from the `tasks` directory
try {
  require('require-dir')('tasks');
} catch (err) {
}
